### DISCUSSION TITLE
Support for tracking database calls

### DISCUSSION THREAD

**Discussion by sandcastle** [JasperFx/marten#1025] [Q&A] [answered]
I'm looking for a solution that would allow me to add a tracking layer (with timings) to all database requests so that I can track them as `Dependencies` for `Application Insights`.

I can't seem to see a path forward at this stage, without wrapping calls in the API individually.

Is there any appetite to add support for something like this? And if so, are there any proposed API's that have been suggested?

#### IDocumentSessionListener

Looking at the `IDocumentSessionListener` interface appears to be a single instance for a store, and I can't see any other mechanism to "tag" a session to correlate the `BeforeSaveChanges` and `AfterCommit` calls. There is also no way of tapping into something like a `BeforeExecute` on load calls.

```csharp
public interface IDocumentSessionListener
{
    void BeforeSaveChanges(IDocumentSession session);
    Task BeforeSaveChangesAsync(IDocumentSession session, CancellationToken token);
    void AfterCommit(IDocumentSession session, IChangeSet commit);
    Task AfterCommitAsync(IDocumentSession session, IChangeSet commit, CancellationToken token);
    void DocumentLoaded(object id, object document);
    void DocumentAddedForStorage(object id, object document);
}
```

#### IMartenSessionLogger

There are also no points in the `IMartenSessionLogger` that would allow for capturing the timing point before and after the command calls on after the call has completed.

```csharp
public interface IMartenSessionLogger
{
    void LogSuccess(NpgsqlCommand command);
    void LogFailure(NpgsqlCommand command, Exception ex);
    void RecordSavedChanges(IDocumentSession session, IChangeSet commit);
}
```

**Comment by jeremydmiller** [comment, marked as answer]
@sandcastle This has come up a couple times. My first suggestion would be to look at the raw tracing in Npgsql itself. 

You *can* build and register an `IDocumentSessionListener` at runtime and add that to a session now (think that was 2.5).
